cdap-app-fabric/src/main/java/io/cdap/cdap/security/executor/ContextInheritingScheduledThreadPoolExecutor.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.executor;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link java.util.concurrent.ScheduledThreadPoolExecutor} which inherits
+ * {@link io.cdap.cdap.security.spi.authentication.SecurityRequestContext}.
+ */
+public class ContextInheritingScheduledThreadPoolExecutor implements ScheduledExecutorService {
```

Comment:
Would you please add some comments on cases in which this logic may not work, as we discussed offline.